<!DOCTYPE html><html><head><meta charSet="utf-8"/><meta name="viewport" content="width=device-width"/><meta name="next-head-count" content="2"/><link rel="preload" href="/_next/static/media/2aaf0723e720e8b9-s.p.woff2" as="font" type="font/woff2" crossorigin="anonymous" data-next-font="size-adjust"/><link rel="preload" href="/_next/static/css/7326cdd8829f50d6.css" as="style"/><link rel="stylesheet" href="/_next/static/css/7326cdd8829f50d6.css" data-n-g=""/><noscript data-n-css=""></noscript><script defer="" nomodule="" src="/_next/static/chunks/polyfills-c67a75d1b6f99dc8.js"></script><script src="/_next/static/chunks/webpack-8fa1640cc84ba8fe.js" defer=""></script><script src="/_next/static/chunks/framework-2c79e2a64abdb08b.js" defer=""></script><script src="/_next/static/chunks/main-b05cd08c4a1aceb2.js" defer=""></script><script src="/_next/static/chunks/pages/_app-f173ab93a2416500.js" defer=""></script><script src="/_next/static/chunks/750-cb383d7bbde68080.js" defer=""></script><script src="/_next/static/chunks/pages/posts/%5Bslug%5D-45676ace6e569333.js" defer=""></script><script src="/_next/static/qBSSL08vI8AZVhEaBf9YA/_buildManifest.js" defer=""></script><script src="/_next/static/qBSSL08vI8AZVhEaBf9YA/_ssgManifest.js" defer=""></script></head><body><div id="__next"><main class="__className_4b5723 text-gray-800"><div class="mx-auto px-4 py-10 max-w-7xl flex justify-between"><a href="/">Home</a></div><div class="mx-auto max-w-7xl px-4 py-8 mb-8 border-b border-slate-100"><div class="text-center text-3xl sm:text-4xl font-bold mb-4">External API error handling</div><div class="text-center mb-4 text-lg text-gray-500">My takeaways from working with shaky APIs</div></div><div class="mx-auto max-w-3xl px-4 prose prose-gray"><div><h2>404 Not found</h2>
<p>First we need to realize that a 404 is not really an error.
The HTTP status is just telling us that the specific resource does not exist.</p>
<pre><span class="hljs-keyword">when</span> (response.status) {
    Ok -&gt; ...
    NotFound -&gt; <span class="hljs-keyword">throw</span> SomethingNotFoundException(...)
}
</pre>
<pre><span class="hljs-keyword">when</span> (response.status) {
    Ok -&gt; ...
    NotFound -&gt; <span class="hljs-literal">null</span>
}
</pre>
<h2>504 Gateway timeout</h2>
<h3>Asynchronous use cases</h3>
<h3>Synchronous use cases</h3>
<h2>500 Internal server error</h2>
<p>Let's talk about the elephant in the room.</p>
<h2>&quot;Unrecoverable&quot; errors</h2>
<h3>400 Bad request</h3>
<p>If your requests are badly formatted it is unlikely that any retry attempt will succeed.</p>
<h3>401 Unauthorized &amp; 403 Forbidden</h3>
<p>Well, you're toast.</p>
<h2>Conclusion</h2>
<ul>
<li>404 is not an error.</li>
<li>504 Gateway timeouts should probably be aborted early and retried</li>
<li>500 internal server errors should be evaluated and retried when possible</li>
</ul>
</div></div><div class="py-8"></div></main></div><script id="__NEXT_DATA__" type="application/json">{"props":{"pageProps":{"post":{"slug":"external-api-error-handling","content":"\n## 404 Not found\n\nFirst we need to realize that a 404 is not really an error.\nThe HTTP status is just telling us that the specific resource does not exist.\n\n```kotlin\nwhen (response.status) {\n    Ok -\u003e ...\n    NotFound -\u003e throw SomethingNotFoundException(...)\n}\n```\n\n```kotlin\nwhen (response.status) {\n    Ok -\u003e ...\n    NotFound -\u003e null\n}\n```\n\n## 504 Gateway timeout\n\n### Asynchronous use cases\n\n### Synchronous use cases\n\n## 500 Internal server error\n\nLet's talk about the elephant in the room.\n\n## \"Unrecoverable\" errors\n\n### 400 Bad request\n\nIf your requests are badly formatted it is unlikely that any retry attempt will succeed.\n\n### 401 Unauthorized \u0026 403 Forbidden\n\nWell, you're toast.\n\n## Conclusion\n\n- 404 is not an error.\n- 504 Gateway timeouts should probably be aborted early and retried\n- 500 internal server errors should be evaluated and retried when possible\n","title":"External API error handling","description":"My takeaways from working with shaky APIs","date":"Jan 16, 2023","draft":true}},"__N_SSG":true},"page":"/posts/[slug]","query":{"slug":"external-api-error-handling"},"buildId":"qBSSL08vI8AZVhEaBf9YA","isFallback":false,"gsp":true,"scriptLoader":[]}</script></body></html>